src/main.rs

```diff
@@ -207,6 +132,35 @@ async fn api_forward_auth(req: Request<IncomingBody>) -> Result<Response<BoxBody
 
 // Login route
 async fn api_login(req: Request<IncomingBody>) -> Result<Response<BoxBody>> {
+    // Get headers from request
+    let headers = req.headers();
+    // Middleware - rate limiter
+    if let Some(mut rate_limiter) = middleware::RateLimiter::global() {
+        // Rate limiter is enabled, grab IP from header and check rate limit
+        if headers.contains_key(FORWARDED_FOR) {
+            let x_forwarded_for = headers[FORWARDED_FOR].to_str().unwrap();
+            let source_ip = x_forwarded_for.split(',').next().unwrap_or(x_forwarded_for);
+            if let Ok(ip) = source_ip.trim().parse::<IpAddr>() {
```

Comment:
That should work perfectly fine for IPv4. Could you use the `/64` subnet instead of the full IPv6 though?

An /64 subnet is the **smallest** subnet that is allowed to be assigned by an internet provider. So average user has access to 2⁶⁴ IPv6 addresses which they can use to circumvent our ratelimiting.

e.g.

```
IPv6: 2a03:2880:f157:83:face:b00c:0:25de
Subnet: 2a03:2880:f157:83::/64
```

That's also how AWS&Co are handling this :wink:

https://adam-p.ca/blog/2022/02/ipv6-rate-limiting/